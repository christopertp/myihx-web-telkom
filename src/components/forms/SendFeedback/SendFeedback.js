import React from 'react';
import { Field } from 'redux-form';
import { useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import Button from '../../../components/elements/Button';
import SelectField from '../../fields/Select';
import TextAreaField from '../../fields/TextArea';
import Spinner from '../../elements/Spinner';
import styles from './styles.scoped.css';

export default function SendFeedback(props) {
  const { handleSubmit, invalid } = props;
  const { isLoading, topic, message } = useSelector(s => s.sendFeedback);
  const inputProps = [
    { label: 'TOPIC', placeholder: isLoading.t ? 'Loading...' : 'Choose...' },
    { label: 'MESSAGE/COMMENT', maxLength: '150', placeholder: 'Enter message' }
  ];

  return (
    <form className={styles.root} onSubmit={handleSubmit}>
      <Field component={SelectField} inputProps={inputProps[0]} name="topicId" options={topic} />
      <Field component={TextAreaField} inputProps={inputProps[1]} name="message" />
      {message && <small className={styles.error}>{message}</small>}
      <Button disabled={invalid || isLoading.s} type="submit">
        {isLoading.s ? <Spinner /> : 'Send Feedback'}
      </Button>
    </form>
  );
}

SendFeedback.defaultProps = {
  handleSubmit: () => { },
  invalid: false,
};

SendFeedback.propTypes = {
  handleSubmit: PropTypes.func,
  invalid: PropTypes.bool,
};
