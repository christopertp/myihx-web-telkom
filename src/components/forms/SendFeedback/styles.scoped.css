.error {
  color: #ee4c24;
  display: block;
  font-size: 0.75rem;
  margin-top: 1rem;
}

.root > button {
  margin-top: 3rem;
  width: 100%;
}

.root > label:nth-of-type(2) {
  margin-top: 1rem;
}

@media only screen and (max-width: 48em) {
  .root {
    margin-top: 5rem;
  }
}