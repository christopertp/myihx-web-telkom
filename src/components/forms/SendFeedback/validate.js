export default function validate(values) {
  const { topicId, message } = values;
  const mustBeFilled = 'Harus diisi';

  return {
    topicId: !topicId ? 'Harus dipilih' : '',
    message: !message ? mustBeFilled : '',
  };
}
