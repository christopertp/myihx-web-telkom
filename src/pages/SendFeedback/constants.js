export const DATA_FETCHED = 'SendFeedback/fetched';
export const FAILED = 'SendFeedback/failed';
export const LOADING = 'SendFeedback/loading';
