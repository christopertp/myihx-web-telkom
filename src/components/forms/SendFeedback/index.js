import { reduxForm } from 'redux-form';
import SendFeedback from './SendFeedback';
import validate from './validate';

export default reduxForm({
  initialValues: {
    topicId: '',
    message: '',
  },
  form: 'sendFeedback',
  validate,
})(SendFeedback);
