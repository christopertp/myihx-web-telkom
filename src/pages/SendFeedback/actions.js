import { feedbackCategory, feedbackSend } from '../../services/nps';
import { push } from 'react-router-redux';
import { DATA_FETCHED, LOADING, FAILED } from './constants';

export function fetchData() {
  return async dispatch => {
    dispatch(loadingAction(true, 'topic'));
    try {
      const { data } = await feedbackCategory();
      dispatch(dataFetchedAction(data));
    } catch (err) {
      dispatch(dataFetchedAction([]));
    }
  };
}

export function fetchSendFeedback(data) {
  return async dispatch => {
    dispatch(loadingAction(true, 'submit'));
    try {
      await feedbackSend(data);
      dispatch(push({ pathname: '/profile', state: { notif: 'Feedback has been send successfully.' } }));
      dispatch(failedAction('', 'submit'));
    } catch (err) {
      dispatch(failedAction('Failed to send data. Please contact developer.', 'submit'));
    }
  };
}

function loadingAction(isLoading, key) {
  return { type: LOADING, isLoading, key };
}

function dataFetchedAction(topic) {
  return { type: DATA_FETCHED, topic };
}

function failedAction(message, key) {
  return { type: FAILED, message, key };
}
