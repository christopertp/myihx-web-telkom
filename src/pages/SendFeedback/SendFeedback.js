import React, { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import SendFeedbackForm from '../../components/forms/SendFeedback';
import WhiteCardPlain from '../../components/layouts/WhiteCardPlain';
import { fetchData, fetchSendFeedback } from './actions';

export default function SendFeedback() {
  const dispatch = useDispatch();
  const onSubmit = payload => {
    dispatch(fetchSendFeedback({ ...payload, apps: 'web' }));
  };

  useEffect(() => {
    dispatch(fetchData());
  }, []);

  const subtitle = `Have any critiques or suggestions
  about our service or the app? We'd love to hear from you!`;

  return (
    <WhiteCardPlain back="/profile" subtitle={subtitle} title="Send Feedback">
      <SendFeedbackForm onSubmit={onSubmit} />
    </WhiteCardPlain>
  );
}
