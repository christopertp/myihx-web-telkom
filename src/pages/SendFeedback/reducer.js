import { fetchData, fetchSendFeedback } from './actions';
import { DATA_FETCHED, FAILED, LOADING } from './constants';

const initialState = {
  fetchData,
  fetchSendFeedback,
  topic: [],
  isLoading: {
    t: true,
    s: false,
  },
  message: '',
};

export default function reducer(state = initialState, action = {}) {
  const { type, topic, isLoading, message, key } = action;
  switch (type) {
    case DATA_FETCHED:
      return {
        ...state,
        topic: topic.map(i => ({ value: i.topicId, text: i.topicName })),
        isLoading: { ...state.isLoading, t: false },
      };
    case FAILED:
      return {
        ...state,
        isLoading: { ...state.isLoading, [key.charAt(0)]: false },
        message,
      };
    case LOADING:
      return {
        ...state,
        isLoading: { ...state.isLoading, [key.charAt(0)]: isLoading },
      };
    default:
      return state;
  }
}
