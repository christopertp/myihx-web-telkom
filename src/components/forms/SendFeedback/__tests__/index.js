import React from 'react';
import { useSelector } from 'react-redux';
import ShallowRenderer from 'react-test-renderer/shallow';
import Spinner from '../../../elements/Spinner';
import SendFeedback from '../SendFeedback';

useSelector.mockImplementation(fn => {
  fn({});
  return {
    isLoading: {
      t: false,
      s: false,
    },
    message: '',
  };
});

describe('src/components/forms/SendFeedback', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<SendFeedback />);
    expect(tree).toMatchSnapshot();
    useSelector.mockImplementationOnce(() => ({
      isLoading: {
        t: true,
        s: true,
      },
      message: 'error.',
    }));
    SendFeedback.defaultProps.handleSubmit();
    const result = SendFeedback({
      ...SendFeedback.defaultProps,
    });
    const button = result.props.children[3];
    expect(button.props.children.type).toBe(Spinner);
  });
});
