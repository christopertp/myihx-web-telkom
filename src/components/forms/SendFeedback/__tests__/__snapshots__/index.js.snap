// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/forms/SendFeedback render 1`] = `
<form
  onSubmit={[Function]}
>
  <Field
    component={[Function]}
    inputProps={
      Object {
        "label": "TOPIC",
        "placeholder": "Choose...",
      }
    }
    name="topicId"
  />
  <Field
    component={[Function]}
    inputProps={
      Object {
        "label": "MESSAGE/COMMENT",
        "maxLength": "150",
        "placeholder": "Enter message",
      }
    }
    name="message"
  />
  
  <Button
    className=""
    disabled={false}
    onClick={null}
    to=""
    type="submit"
    variant="red"
  >
    Send Feedback
  </Button>
</form>
`;
